### ISSUE TITLE
Loss of VLM fine-tuning

### ISSUE THREAD

**Issue by shashacks** [DigiRL-agent/digiq#14]
I mostly followed Section 4.
I set eval_iterations to 81 to collect around 1200 trajectories, as mentioned in your paper. (I wasn't able to get 1296 due to a Gemini API call error.)

I fine-tuned LLaVA through:
cd <path_to_digiq>/postprocess_data/vlm_finetune/
bash finetune.sh

I changed num_train_epochs from 0.01 to 3 (I checked the comment: Training for more than 2 epochs is usually enough.)

However, the losses are always 0 after the first epoch, as shown below. Could you verify this?

If the loss is 0, I'm not sure why we need to train for up to 3 epochs.

{'loss': 0.2528, 'learning_rate': 3.3333333333333333e-06, 'epoch': 0.02}                                                                                                                                                  
{'loss': 0.0, 'learning_rate': 6.666666666666667e-06, 'epoch': 0.03}                                                                                                                                                      
{'loss': 0.0, 'learning_rate': 1e-05, 'epoch': 0.05}                                                                                                                                                                      
{'loss': 0.0, 'learning_rate': 1.3333333333333333e-05, 'epoch': 0.07}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.6666666666666667e-05, 'epoch': 0.09}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 2e-05, 'epoch': 0.1}                                                                                                                                                                       
{'loss': 0.0, 'learning_rate': 1.9998251609127465e-05, 'epoch': 0.12}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.9993007047883988e-05, 'epoch': 0.14}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.998426815017817e-05, 'epoch': 0.16}                                                                                                                                                      
{'loss': 0.0, 'learning_rate': 1.9972037971811802e-05, 'epoch': 0.17}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.9956320789411338e-05, 'epoch': 0.19}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.9937122098932428e-05, 'epoch': 0.21}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.9914448613738107e-05, 'epoch': 0.22}                                                                                                                                                     
{'loss': 0.0, 'learning_rate': 1.9888308262251286e-05, 'epoch': 0.24}

**Comment by BiEchi** [comment]
Your loss is far lower than what we observed with the released data. Would you like to try using our released data for training LLaVA? To do this, just use the released data to generate the contrastive datapoints by modifying `input_data_path` below to our released data:

```bash
llava_training_datapoints_path=$working_dir/datapoints-llava-contrastive.json
python -u construct_sft_data.py \
    --input_data_path $layover_image_data_path \
    --output_contrastive_data_path $llava_training_datapoints_path
```

If this results in a loss starting from 5 to 0.25, then your pipeline looks good, from where you can use your own data to resume.

**Comment by shashacks** [comment]
I will also try it, but what I noticed is that there's a huge difference in file size between the two: the file I generated—trajectories_eval.pt (which I generated twice in separate folders to double-check)—is around 186MB, while the provided data, autoui-zeroshot-trajectories/general-1008trajs.pt, is 6.71GB. However, the image sizes in both cases are similar (around 3GB). I'm curious why general-1008trajs.pt is so much larger than the image folder, given that it only includes text and vision features—Maybe it should not be larger than the image folder?..

**Comment by BiEchi** [comment]
@shashacks this is a good question. The reason is that the file `autoui-zeroshot-trajectories/general-1008trajs.pt` contains the image representations of all the actions (64 actions in total, each action will layover the image, so in total 64 images), so it is 64x larger than a file with single representations.

**Comment by BiEchi** [comment]
Note however, to obtain the datapoints for contrastive learning, you don't need all of these actions (you only need the original action), so the size of your collected trajectories looks fine to me as well (around 1/64).

**Comment by shashacks** [comment]
Hmm, it's a bit confusing. Could you clarify which commands I should follow in get_more_actions/RUNME.sh when using general-1008trajs.pt? If general-1008trajs.pt already includes the additional action features, then maybe we don't need to do Step 4 (collect more actions). But we may still need the smaller images and cursor images? Should I follow Steps 1 to 6, except Step 4?

Also, are the q-image paths already linked in general-1008trajs.pt?
And finally, should we change num_actions_total to 64?

**Comment by BiEchi** [comment]
As mentioned in my previous response, you can simply run [this command](https://github.com/DigiRL-agent/digiq/blob/8fbcfa4fbf83af7376b5097d86a359f40af25239/postprocess_data/get_more_actions/RUNME.sh#L54):

```
llava_training_datapoints_path=$working_dir/datapoints-llava-contrastive.json
python -u construct_sft_data.py \
    --input_data_path $layover_image_data_path \
    --output_contrastive_data_path $llava_training_datapoints_path
```

Just replace $layover_image_data_path with general-1008trajs.pt and other paths are up to you.

**Comment by BiEchi** [comment]
For your other questions:

1. Yes, the Q image paths are already linked in general-1008trajs.pt, and you can find them in the image folders we released.
2. You don't need to run other commands in the script so no need to set `num_actions_total` at all.

**Comment by shashacks** [comment]
I printed the keys each step contains, and there’s no next_image_path, q_image_path, step_id, or traj_id, which makes it difficult to run construct_sft_data.py as-is.

It would be great if you could try running it on your end to confirm what stages are necessary and which can be skipped. From what I can tell, steps 1 and 2 are probably required (with some modifications), especially in nv2v.py, to correctly link an image to the next image.

dict_keys(['observation', 'next_observation', 'image_features', 'image_path', 'next_image_features', 'task', 'reward', 'done', 'action', 'trajectory_reward', 'mc_return', 'action_list', 'q_rep_out_list', 'q_rep_out'])
<pre>  
for traj_id, traj in tqdm(enumerate(trajs)):
    for step_id, step in enumerate(traj):
        print(step.keys())
        exit()
        datapoint = {
            "id": f"{step['traj_id']}-{step['step_id']}",
            "image": step["q_image_path"].split("/")[-1],
            "conversations": [
            {"from": "human", "value": make_prompt(step['parsed_action'])},
            {"from": "gpt", "value": judge(step["image_path"], step["next_image_path"])}
 </pre>

**Comment by Youngjae-KAIST** [comment]
I am having similar problem on both custom and your trajectory files.

{'loss': 0.2356, 'grad_norm': 0.0015397267572645774, 'learning_rate': 1.8e-05, 'epoch': 0.1}                                                                                                                                          
{'loss': 0.0, 'grad_norm': 0.0020372979860289685, 'learning_rate': 1.9955322512672162e-05, 'epoch': 0.19}                                                                                                                             
{'loss': 0.0, 'grad_norm': 0.00031518082713659904, 'learning_rate': 1.980139427847242e-05, 'epoch': 0.29}                                                                                                                             
{'loss': 0.0, 'grad_norm': 0.00024456712018669827, 'learning_rate': 1.953936107650336e-05, 'epoch': 0.39}                                                                                                                             
{'loss': 0.0, 'grad_norm': 0.00014876364074331387, 'learning_rate': 1.917211301505453e-05, 'epoch': 0.49}                                                                                                                             
{'loss': 0.0, 'grad_norm': 9.702665263115864e-05, 'learning_rate': 1.870370067496355e-05, 'epoch': 0.58}                                                                                                                              
{'loss': 0.0, 'grad_norm': 6.885081945230464e-05, 'learning_rate': 1.8139290433532415e-05, 'epoch': 0.68}                                                                                                                             
{'loss': 0.0, 'grad_norm': 4.7397717987640554e-05, 'learning_rate': 1.7485107481711014e-05, 'epoch': 0.78}

Seems like latest flash_attn does not support torch2.1.2 which LLaVa is using.

https://github.com/Dao-AILab/flash-attention/releases

I had to downgrade to flash_attn v2.5.8 which looks stable.
Then, accelerate, deepspeed, transformers and all dependency was broken.
Upgrading all of them solves the problem but had to modify LLaVa codebase.

Could you please tell us the version of flash_attn you used?

I can then rerun the code but it would be better if you could verify it.
Just in case where package version is not a problem.

**Comment by BiEchi** [comment]
@shashacks You don't need to run `nv2v.py`, it's only for the raw data you collected, running it on an already processed data (like the released data) will result in offset errors. **You simply need to generate Q images** with `generate_more_click_images_batched.py`. The next image can be obtained just by accessing the image path of the next step in the same trajectory, i.e. with `traj[i][j+1]['image_path']`. The traj_id and step_id does not matter at all! You can just randomly assign then as long as they don't repeat.

In case you still feel it a bit too complicated to reproduce with the original data, you can simply use your own data for fine-tuning. In our practice, if you observe most of the time the yes/no output makes sense, you're good to go, it does not matter how many epochs you tune if you observe the output makes sense.

**Comment by BiEchi** [comment]
@Youngjae-KAIST As mentioned in README, you should probably create a new conda environment for finetuning & generating the llava representations. After that, just follow the standard setup shown by the [LLaVA repo](https://github.com/haotian-liu/LLaVA).

**Comment by BiEchi** [comment]
Also @Youngjae-KAIST would you like to show the commands you used to train the llava representations with the released data? This will be very helpful for @shashacks to catch up and debug.

**Comment by BiEchi** [comment]
@shashacks @Youngjae-KAIST 

Actually I suspect the targets you guys obtained might be all "Yes"s or all "No"s - would you like to check the ratio of the datapoints generated? The released data has a roughly 0.4:0.6 ratio. Although this ratio does not matter that much (as shown in our [rebuttal](https://openreview.net/forum?id=CjfQssZtAb)), you should definitely check whether the ratio gets ridiculous high or low.

**Comment by Youngjae-KAIST** [comment]
Thank you for your answer.

I realized that targets are always "no".
They were pointing to same images.
I accidently changed positions of code when factorizing.

It now has ratio of 0.6 and reaches 0.8 at the end

**Comment by BiEchi** [comment]
Nice! cc'ing @shashacks

**Comment by Youngjae-KAIST** [comment]
>would you like to show the commands you used to train the llava representations with the released data? 

@shashacks I just followed the whole code in postprocess_data/get_more_actions
each of them are kind of depedendent on precedent ouput.
we need, for example, cursor_images folder to be generated before finetuning.

However, to test on [the given data](https://huggingface.co/datasets/JackBAI/autoui-zeroshot-trajectories/tree/main). You should remove special tokens.
Adding a line suffice for this

```
image_path_without_special_tokens =  re.sub(r".*?/<path_to_images>/?", "", orig_trajs[i][j]['image_path'])
```

I think you're also having similar problem on your data. so I suppose checking the data would be better
